File: backend/src/routes/labtests.js
import { Router } from 'express';
import { z } from 'zod';
import { requireAuth, resolvePatientScope } from '../middleware/auth.js';
import { validate } from '../middleware/validate.js';
import { asyncHandler } from '../middleware/errors.js';
import { audit } from '../middleware/audit.js';
import { LabResult } from '../models/LabResult.js';
import { Prescription } from '../models/Prescription.js';

/**
 * The patient's lab tests: the tests the doctor advised (pulled from active
 * prescriptions) and the reports the patient has uploaded against them.
 * Mounted at /patients/:patientId/lab-tests.
 */
const router = Router({ mergeParams: true });
router.use(requireAuth, resolvePatientScope);

function serialiseResult(r) {
  return {
    id: String(r._id),
    testName: r.testName,
    note: r.note ?? '',
    photoUrl: r.photo ? `/api/v1/uploads/${r.photo}/raw` : null,
    createdAt: r.createdAt,
  };
}

router.get(
  '/',
  audit('read', 'LabResult'),
  asyncHandler(async (req, res) => {
    const [prescriptions, results] = await Promise.all([
      Prescription.find({ patient: req.patientId, isActive: true }).select('labTestsAdvised').lean(),
      LabResult.find({ patient: req.patientId }).sort({ createdAt: -1 }).limit(100).lean(),
    ]);
    const advised = [...new Set(prescriptions.flatMap((p) => p.labTestsAdvised ?? []).filter(Boolean))];
    res.json({ advised, results: results.map(serialiseResult) });
  }),
);

router.post(
  '/',
  validate({
    body: z
      .object({
        testName: z.string().trim().min(1).max(200),
        note: z.string().trim().max(1000).optional().default(''),
        photo: z.string().optional(),
      })
      .refine((b) => Boolean(b.photo) || b.note.trim().length > 0, { message: 'Add a photo or a note', path: ['photo'] }),
  }),
  audit('create', 'LabResult'),
  asyncHandler(async (req, res) => {
    const entry = await LabResult.create({
      patient: req.patientId,
      testName: req.body.testName,
      note: req.body.note,
      photo: req.body.photo || undefined,
    });
    res.status(201).json({ result: serialiseResult(entry) });
  }),
);

export default router;

File: mobile/lib/features/labtests/domain/lab_tests.dart
/// A test report the patient uploaded against a doctor-advised test.
class LabResult {
  const LabResult({required this.id, required this.testName, required this.note, this.photoUrl, this.createdAt});

  final String id;
  final String testName;
  final String note;
  final String? photoUrl;
  final DateTime? createdAt;

  factory LabResult.fromJson(Map<String, dynamic> j) => LabResult(
        id: j['id']?.toString() ?? '',
        testName: j['testName']?.toString() ?? '',
        note: j['note']?.toString() ?? '',
        photoUrl: (j['photoUrl'] == null || j['photoUrl'].toString().isEmpty) ? null : j['photoUrl'].toString(),
        createdAt: DateTime.tryParse(j['createdAt']?.toString() ?? '')?.toLocal(),
      );
}

/// The tests the doctor advised + the reports the patient has uploaded.
class LabTestsView {
  const LabTestsView({required this.advised, required this.results});

  final List<String> advised;
  final List<LabResult> results;

  /// True once a report has been uploaded for [test].
  bool hasResultFor(String test) => results.any((r) => r.testName.toLowerCase() == test.toLowerCase());

  factory LabTestsView.fromJson(Map<String, dynamic> j) => LabTestsView(
        advised: (j['advised'] as List?)?.map((e) => e.toString()).toList() ?? const [],
        results: (j['results'] as List?)?.whereType<Map<String, dynamic>>().map(LabResult.fromJson).toList() ?? const [],
      );
}
